File: src/main/java/io/github/chindeaone/collectiontracker/utils/SkillUtils.java
/*
    Calculations based on Skyhanni's logic
 */
package io.github.chindeaone.collectiontracker.utils;

import java.util.Arrays;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class SkillUtils {

    public enum Skills {
        MINING("Mining"),
        FORAGING("Foraging"),
        FISHING("Fishing"),
        FARMING("Farming"),
        COMBAT("Combat"),
        ALCHEMY("Alchemy"),
        ENCHANTING("Enchanting"),
        TAMING("Taming"),
        HUNTING("Hunting"),
        CARPENTRY("Carpentry"),
        RUNECRAFTING("Runecrafting"),
        SOCIAL("Social");

        private final String displayName;

        Skills(String displayName) {
            this.displayName = displayName;
        }

        public String getDisplayName() {
            return displayName;
        }
    }

    public enum MaxSkillLevels {
        MINING(60),
        FORAGING(54),
        FISHING(50),
        FARMING(60),
        COMBAT(60),
        ALCHEMY(50),
        ENCHANTING(60),
        TAMING(60),
        HUNTING(25),
        CARPENTRY(50),
        RUNECRAFTING(25),
        SOCIAL(25);

        private final int maxLevel;

        MaxSkillLevels(int maxLevel) {
            this.maxLevel = maxLevel;
        }

        public int getMaxLevel() {
            return maxLevel;
        }
    }

    public static final EnumMap<Skills, Double> SKILL_VALUES = new EnumMap<>(Skills.class);
    public static final EnumMap<Skills, Integer> SKILL_LEVELS = new EnumMap<>(Skills.class);
    public static final EnumMap<Skills, Long> MAX_SKILL_XP = new EnumMap<>(Skills.class);

    private static final double XP_LVL_60 = 111_672_425D;

    // Precompute max XP for each skill
    public static void initializeMaxSkillXp() {
        for (Skills skill : Skills.values()) {
            int maxLevel = getMaxLevelForSkill(skill);
            long maxXp = getTotalXpForMaxLevel(skill, maxLevel);
            MAX_SKILL_XP.put(skill, maxXp);
        }
    }

    public static List<String> getDisplayNames() {
        return Arrays.stream(Skills.values())
                .map(Skills::getDisplayName)
                .collect(Collectors.toList());
    }

    public static boolean isValidSkill(String skillName) {
        return getDisplayNames().contains(skillName);
    }

    public static void updateFromApi(Map<String, Double> apiValues) {
        if (apiValues == null || apiValues.isEmpty()) return;
        synchronized (SKILL_VALUES) {
            synchronized (SKILL_LEVELS) {
                for (Map.Entry<String, Double> entry : apiValues.entrySet()) {
                    Skills skill = fromDisplayName(entry.getKey());
                    Double value = entry.getValue();
                    if (skill != null && value != null) {
                        int level = getLevelFromXp(skill, value);
                        SKILL_VALUES.put(skill, value);
                        SKILL_LEVELS.put(skill, level);
                    }
                }
            }
        }
        initializeMaxSkillXp();
    }

    public static Double getSkillValue(String skillName) {
        Skills skill = fromDisplayName(skillName);
        return skill == null ? null : getSkillValue(skill);
    }

    public static Integer getSkillLevel(String skillName) {
        Skills skill = fromDisplayName(skillName);
        return skill == null ? null : getSkillLevel(skill);
    }

    public static Double getTamingValue() {
        return getSkillValue(Skills.TAMING);
    }

    public static Integer getTamingLevel() {
        return getSkillLevel(Skills.TAMING);
    }

    public static long getMaxXpForSkill(String skillName) {
        Skills skill = fromDisplayName(skillName);
        return skill == null ? 0L : MAX_SKILL_XP.get(skill);
    }

    private static long getTotalXpForMaxLevel(Skills skill, int level) {
        if (level <=0) return 0L;
        if (skill == Skills.RUNECRAFTING) {
            return getCumulativeFromTable(RUNECRAFTING_LEVELING_XP, level);
        }
        if (level <=60) {
            return getCumulativeFromTable(LEVELING_XP, level);
        }
        return 0L;
    }

    private static long getCumulativeFromTable(int[] table, int level) {
        long total =0L;
        int capped = Math.min(level, table.length);
        for (int i =0; i < capped; i++) {
            total += table[i];
        }
        return total;
    }

    public static Boolean isSkillMaxed(String skillName) {
        Skills skill = toSkill(skillName);
        return skill == null ? null : isSkillMaxed(skill);
    }

    private static Boolean isSkillMaxed(Skills skill) {
        synchronized (SKILL_LEVELS) {
            Integer level = SKILL_LEVELS.get(skill);
            return level != null && level >= getMaxLevelForSkill(skill);
        }
    }

    private static Double getSkillValue(Skills skill) {
        synchronized (SKILL_VALUES) {
            return SKILL_VALUES.get(skill);
        }
    }

    private static Integer getSkillLevel(Skills skill) {
        synchronized (SKILL_LEVELS) {
            return SKILL_LEVELS.get(skill);
        }
    }

    private static Skills toSkill(String skillName) {
        return fromDisplayName(skillName);
    }

    private static Skills fromDisplayName(String skillName) {
        if (skillName == null) return null;
        for (Skills skill : Skills.values()) {
            if (skill.getDisplayName().equalsIgnoreCase(skillName)) {
                return skill;
            }
        }
        return null;
    }

    private static int getMaxLevelForSkill(Skills skill) {
        return MaxSkillLevels.valueOf(skill.name()).getMaxLevel();
    }

    private static int getLevelFromXp(Skills skill, double totalXp) {
        if (skill == Skills.RUNECRAFTING) {
            return getLevelFromTable(RUNECRAFTING_LEVELING_XP, totalXp);
        }
        if (totalXp < XP_LVL_60) {
            return getLevelFromTable(LEVELING_XP, totalXp);
        }
        return getLevelAbove60(totalXp);
    }

    private static int getLevelFromTable(int[] table, double totalXp) {
        long xp = (long) Math.floor(totalXp);
        long cumulative = 0L;
        for (int i = 0; i < table.length; i++) {
            cumulative += table[i];
            if (xp < cumulative) {
                return i;
            }
        }
        return table.length;
    }

    private static int getLevelAbove60(double totalXp) {
        long remaining = (long) Math.floor(totalXp - XP_LVL_60);
        int level = 60;
        while (remaining >= xpForLevelAbove60(level + 1)) {
            remaining -= xpForLevelAbove60(level + 1);
            level++;
        }
        return level;
    }

    private static long xpForLevelAbove60(int level) {
        if (level <= 60) return 0L;
        int k = level - 60; // 1 for lvl 61
        int block = (k - 1) / 10; // which 10-level block
        int posInBlock = (k - 1) % 10; // index inside the block (0-9)
        long baseXp = 7_600_000L;
        long baseSlope = 600_000L;
        long slope = baseSlope << block; // slope * 2^block
        long blockStartXp = baseXp + (10L * baseSlope * ((1L << block) - 1));
        return blockStartXp + slope * posInBlock;
    }

    private static final int[] LEVELING_XP = new int[] {
            50, 125, 200, 300, 500, 750, 1000, 1500, 2000, 3500,
            5000, 7500, 10000, 15000, 20000, 30000, 50000, 75000, 100000, 200000,
            300000, 400000, 500000, 600000, 700000, 800000, 900000, 1000000, 1100000, 1200000,
            1300000, 1400000, 1500000, 1600000, 1700000, 1800000, 1900000, 2000000, 2100000, 2200000,
            2300000, 2400000, 2500000, 2600000, 2750000, 2900000, 3100000, 3400000, 3700000, 4000000,
            4300000, 4600000, 4900000, 5200000, 5500000, 5800000, 6100000, 6400000, 6700000, 7000000
    };

    private static final int [] RUNECRAFTING_LEVELING_XP = new int[] {
            50, 100, 125, 160, 200, 250, 315, 400, 500, 625,
            785, 1000, 1250, 1600, 2000, 2465, 3125, 4000, 5000, 6200,
            7800, 9800, 12200, 15300, 19050
    };
}

File: src/main/java/io/github/chindeaone/collectiontracker/utils/SkillUtils.kt
/*
    Calculations based on Skyhanni's logic
 */
package io.github.chindeaone.collectiontracker.utils

import java.util.EnumMap
import kotlin.math.floor
import kotlin.math.min

object SkillUtils {
    val SKILL_VALUES = EnumMap<Skills, Double>(Skills::class.java)
    val SKILL_LEVELS = EnumMap<Skills, Int>(Skills::class.java)
    val MAX_SKILL_XP = EnumMap<Skills, Long>(Skills::class.java)

    private const val XP_LVL_60 = 111_672_425.0

    // Precompute max XP for each skill
    fun initializeMaxSkillXp() {
        for (skill in Skills.entries) {
            val maxLevel = getMaxLevelForSkill(skill)
            val maxXp = getTotalXpForMaxLevel(skill, maxLevel)
            MAX_SKILL_XP[skill] = maxXp
        }
    }

    fun getDisplayNames(): List<String> = Skills.entries.map { it.displayName }

    fun isValidSkill(skillName: String): Boolean = Skills.entries.any { it.displayName.equals(skillName, ignoreCase = true) }

    fun updateFromApi(apiValues: Map<String, Double>) {
        if (apiValues.isEmpty()) return
        synchronized(SKILL_VALUES) {
            synchronized(SKILL_LEVELS) {
                for (entry in apiValues.entries) {
                    val skill = fromDisplayName(entry.key)
                    val value = entry.value
                    if (skill != null) {
                        val level = getLevelFromXp(skill, value)
                        SKILL_VALUES[skill] = value
                        SKILL_LEVELS[skill] = level
                    }
                }
            }
        }
        initializeMaxSkillXp()
    }

    fun getSkillValue(skillName: String): Double? = getSkillValue(fromDisplayName(skillName) ?: return null)

    fun getSkillLevel(skillName: String): Int? = getSkillLevel(fromDisplayName(skillName) ?: return null)

    val tamingValue: Double get() = getSkillValue(Skills.TAMING)

    val tamingLevel: Int get() = getSkillLevel(Skills.TAMING)

    fun getMaxXpForSkill(skillName: String): Long {
        val skill = fromDisplayName(skillName)
        return skill?.let { MAX_SKILL_XP[it] ?: 0L } ?: 0L
    }

    private fun getTotalXpForMaxLevel(skill: Skills?, level: Int): Long {
        if (level <= 0) return 0L
        if (skill == Skills.RUNECRAFTING) {
            return getCumulativeFromTable(RUNECRAFTING_LEVELING_XP, level)
        }
        if (level <= 60) {
            return getCumulativeFromTable(LEVELING_XP, level)
        }
        return 0L
    }

    private fun getCumulativeFromTable(table: IntArray, level: Int): Long {
        var total = 0L
        val capped = min(level, table.size)
        for (i in 0..<capped) {
            total += table[i].toLong()
        }
        return total
    }

    fun isSkillMaxed(skillName: String): Boolean = isSkillMaxed(toSkill(skillName) ?: return false)

    private fun isSkillMaxed(skill: Skills): Boolean {
        synchronized(SKILL_LEVELS) {
            val level = SKILL_LEVELS[skill]
            return level != null && level >= getMaxLevelForSkill(skill)
        }
    }

    private fun getSkillValue(skill: Skills): Double = synchronized(SKILL_VALUES) {
        SKILL_VALUES[skill] ?: 0.0
    }

    private fun getSkillLevel(skill: Skills): Int = synchronized(SKILL_LEVELS) {
        SKILL_LEVELS[skill] ?: 0
    }

    private fun toSkill(skillName: String): Skills? = fromDisplayName(skillName)

    private fun fromDisplayName(skillName: String): Skills? = Skills.entries.firstOrNull { it.displayName.equals(skillName, ignoreCase = true) }

    private fun getMaxLevelForSkill(skill: Skills): Int = MaxSkillLevels.valueOf(skill.name).maxLevel

    private fun getLevelFromXp(skill: Skills?, totalXp: Double): Int {
        if (skill == Skills.RUNECRAFTING) {
            return getLevelFromTable(RUNECRAFTING_LEVELING_XP, totalXp)
        }
        if (totalXp < XP_LVL_60) {
            return getLevelFromTable(LEVELING_XP, totalXp)
        }
        return getLevelAbove60(totalXp)
    }

    private fun getLevelFromTable(table: IntArray, totalXp: Double): Int {
        val xp = floor(totalXp).toLong()
        var cumulative = 0L
        for (i in table.indices) {
            cumulative += table[i].toLong()
            if (xp < cumulative) {
                return i
            }
        }
        return table.size
    }

    private fun getLevelAbove60(totalXp: Double): Int {
        var remaining = floor(totalXp - XP_LVL_60).toLong()
        var level = 60
        while (true) {
            val xpNeeded = xpForLevelAbove60(level + 1)
            if (remaining < xpNeeded) break
            remaining -= xpNeeded
            level++
        }
        return level
    }

    private fun xpForLevelAbove60(level: Int): Long {
        if (level <= 60) return 0L
        val k = level - 60 // 1 for lvl 61
        val block = (k - 1) / 10 // which 10-level block
        val posInBlock = (k - 1) % 10 // index inside the block (0-9)
        val baseXp = 7600000L
        val baseSlope = 600000L
        val slope = baseSlope shl block // slope * 2^block
        val blockStartXp = baseXp + (10L * baseSlope * ((1L shl block) - 1))
        return blockStartXp + slope * posInBlock
    }

    private val LEVELING_XP = intArrayOf(
        50, 125, 200, 300, 500, 750, 1000, 1500, 2000, 3500,
        5000, 7500, 10000, 15000, 20000, 30000, 50000, 75000, 100000, 200000,
        300000, 400000, 500000, 600000, 700000, 800000, 900000, 1000000, 1100000, 1200000,
        1300000, 1400000, 1500000, 1600000, 1700000, 1800000, 1900000, 2000000, 2100000, 2200000,
        2300000, 2400000, 2500000, 2600000, 2750000, 2900000, 3100000, 3400000, 3700000, 4000000,
        4300000, 4600000, 4900000, 5200000, 5500000, 5800000, 6100000, 6400000, 6700000, 7000000
    )

    private val RUNECRAFTING_LEVELING_XP = intArrayOf(
        50, 100, 125, 160, 200, 250, 315, 400, 500, 625,
        785, 1000, 1250, 1600, 2000, 2465, 3125, 4000, 5000, 6200,
        7800, 9800, 12200, 15300, 19050
    )

    enum class Skills(val displayName: String) {
        MINING("Mining"),
        FORAGING("Foraging"),
        FISHING("Fishing"),
        FARMING("Farming"),
        COMBAT("Combat"),
        ALCHEMY("Alchemy"),
        ENCHANTING("Enchanting"),
        TAMING("Taming"),
        HUNTING("Hunting"),
        CARPENTRY("Carpentry"),
        RUNECRAFTING("Runecrafting"),
        SOCIAL("Social");

        override fun toString(): String {
            return displayName
        }
    }

    enum class MaxSkillLevels(val maxLevel: Int) {
        MINING(60),
        FORAGING(57),
        FISHING(50),
        FARMING(60),
        COMBAT(60),
        ALCHEMY(50),
        ENCHANTING(60),
        TAMING(60),
        HUNTING(50),
        CARPENTRY(50),
        RUNECRAFTING(25),
        SOCIAL(25)
    }
}